pallets/poa/src/migrations.rs

```diff
@@ -0,0 +1,131 @@
+/*
+ * This file is part of the Nodle Chain distributed at https://github.com/NodleCode/chain
+ * Copyright (C) 2020-2022  Nodle International
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+pub mod v1 {
+	use crate::{Config, Releases, StorageVersion};
+	use frame_support::{
+		pallet_prelude::PhantomData,
+		storage::migration::{have_storage_value, remove_storage_prefix},
+		traits::{Get, OnRuntimeUpgrade},
+		weights::Weight,
+	};
+
+	pub struct MigrateToBoundedValidators<T>(PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToBoundedValidators<T> {
+		fn on_runtime_upgrade() -> Weight {
+			log::info!(
+				"on_runtime_upgrade[{:#?}]=> Running migration with current storage version {:?} / onchain {:?}",
+				line!(),
+				crate::Releases::V2_0_21,
+				<StorageVersion<T>>::get(),
+			);
+
+			if <StorageVersion<T>>::get() == Releases::V0_0_0Legacy {
+				let pallet_prefix: &[u8] = b"Poa";
+				let storage_item_prefix: &[u8] = b"Validators";
+
+				if have_storage_value(pallet_prefix, storage_item_prefix, &[]) {
+					remove_storage_prefix(pallet_prefix, storage_item_prefix, &[]);
+
+					<StorageVersion<T>>::put(crate::Releases::V2_0_21);
+
+					log::info!(
+						"on_runtime_upgrade[{:#?}]=> Removed Validators, Migrated to storage version {:?}",
+						line!(),
+						<StorageVersion<T>>::get()
+					);
+				} else {
+					panic!("on_runtime_upgrade[{:#?}]=> Validators doesn't exist", line!());
```

Comment:
We should test the situation in pre-upgrade and flag it in try-runtime tests. However during the runtime upgrade we should simply try to remove the storage without checking have_storage_value. The reason is panic is not expected at all for our nodes under any circumstances that we can control.

Reply:
Done